Lab1_Operationalizing_Pytorch_with_Mlflow/README.md

```diff
@@ -1,31 +1,126 @@
+| 2 | Durable Data Storage | Every serious solution needs to store durable data | Tracking Server | Google Cloud SQL | A Postgres Database hosted by Google Cloud SQL is reliable and fast enough for our needs  |
+| 3 | Large and Binary Data Storage | ML projects eventually need to store models and extra large files, This type of data doesn't fit in RDBMS | Tracking Server<br>Artifact Store | Google Cloud Storage | Same as #1 |
+| 4 | Shipment and Deployment | There are many deployable assets as well as acessory tools that need specific environment and SO libraries to run. Just installing them into a machine isn't a viable option | Training  Node<br>Google Cloud Services | Docker | It's the _de facto_ pattern for shipping and deploying things |
+| 5 | Lib Management | Python has a powerful but sometimes confusing and conflicting library ecossystem. Our components may depend on conflicting libs and that will lead to problems in Production | All Nodes and Machines | Pip | Pip is heavily used by Python community. It also work quite well with isolation provided by Docker images |
+| 6 | Remote Access | Sometimes we need to issue remote commands or operate a remote machine | All Nodes and Machines | SSH | Hey... It's SSH. There's no need to rationale :-P |
+| 7 | CI/CD & Workflow Management | We need to seamlessly integrate and deploy ML stuff. We also have to coordinate the process between nodes | All Items | Github Actions | We're using GitHub, it's heavily used by community and comes off-the-shelf |
+| 8 | Project Metadata | We need to describe the ML Component, its structure and how it should run | Source Code Repository<br>Training Node | MLFlow Project Definition | Non-functional Requirement we must adhere |
+| 9 | ML Lifecycle Management | We must track the ML Process | All Nodes and Machines | MLFlow |  Non-functional Requirement we must adhere |
+| 10 | ML Framework | We are building ML Components. Therefore, we must use a ML framework to train our models | Training Node<br>Serving Instances | Pytorch | Non-functional Requirement we must adhere |
+
+
+### Workflow
+
+[DevOps](https://en.wikipedia.org/wiki/DevOps) tools heavily relies on actions triggered by changes
+and other events regarding application's source code. MLOps goes beyond this assumption since any reasonable
+solution must deal with assets like ML Models, datasets, features, etc. Therefore, a simples ML 
+application may turn into a complex workflow in order to ensure the best model runs in Production. The
+following workflow depicts a common development scenario where a production web application uses a ML
+Model to provide business value. 
+
+![Solution Workflow](imgs/workflow.svg)
+
+The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
+changes into a Development Server. Here follows the steps:
```

Comment:
Does it makes sense here or before here to talk about the experimentation work that can be done in MLflow before we get to the point of productionisation? I only mention it since the strong part of MLflow is that a (potentially cowboy!) data scientist can do all their experimentation work in MLflow, track parameters, compare the performance etc. Then they decide what to commit to master, and the MLops pipeline we've built out here takes over, but lets them see the progress in their "own" tool.
```suggestion
Before this MLops workflow is used, there would be an experimentation workflow used by a data scientist. They would be able to use this repository to test out various models, architectures, hyperparameters etc. which would all be logged and tracked in MLflow available for them to compare the performance in the UI. Once they had decided on a new model to move to production, they can create a Pull Request, ready for their team to review, which would then move into the MLops workflow proper.
The first stage is Development Workflow. It concerns the continuous integration of new ML code and Dataset 
changes into a Development Server. Here follows the steps:
```

Reply:
John... you're totally right. It's a great suggestion. I thought it would be nice if we could also mention this during the context and give MLFlow a nice spot. So, I did quite a big change in the text. Don't know if its too formal for our purposes. Anyway, take a look and if you think it doesn't fit there we may rollback it and accept your text right away. :wink: